src/services/Storage.tsx

```diff
@@ -0,0 +1,180 @@
+import AsyncStorage from '@react-native-async-storage/async-storage'
+import React, { createContext, ReactElement } from 'react'
+
+import { Favorite, Progress } from '../constants/data'
+import { UserVocabularyItem } from '../constants/endpoints'
+import useLoadAsync from '../hooks/useLoadAsync'
+import { WordNodeCard } from './RepetitionService'
+import { CMS } from './axios'
+import { migrateToNewFavoriteFormat } from './storageUtils'
+
+export type Storage = {
+  wordNodeCards: WordNodeCard[]
+  isTrackingEnabled: boolean
+  // Null means the selected professions were never set before, which means that the intro should be shown
+  selectedProfessions: number[] | null
+  isDevModeEnabled: boolean
+  progress: Progress
+  cmsUrlOverwrite: CMS | null
+  customDisciplines: string[]
+  userVocabulary: UserVocabularyItem[]
+  nextUserVocabularyId: number
+  favorites: Favorite[]
+}
+
+/**
+ * The default storage contains the default values that will be used if nothing else
+ * is stored yet on the device.
+ * It is also useful for testing, to mock the actual storage implementation.
+ */
+export const newDefaultStorage = (): Storage => ({
+  wordNodeCards: [],
+  isTrackingEnabled: true,
+  selectedProfessions: null,
+  isDevModeEnabled: false,
+  progress: {},
+  cmsUrlOverwrite: null,
+  customDisciplines: [],
+  userVocabulary: [],
+  nextUserVocabularyId: 1,
+  favorites: [],
+})
+const defaultStorage = newDefaultStorage()
+
+// eslint-disable-next-line consistent-return
+export const getStorageKey = (key: keyof Storage): string => {
+  switch (key) {
+    case 'wordNodeCards':
+      return 'wordNodeCards'
+    case 'isTrackingEnabled':
+      return 'sentryTracking'
+    case 'selectedProfessions':
+      return 'selectedProfessions'
+    case 'isDevModeEnabled':
+      return 'devmode'
+    case 'progress':
+      return 'progress'
+    case 'cmsUrlOverwrite':
+      return 'cms'
+    case 'customDisciplines':
+      return 'customDisciplines'
+    case 'userVocabulary':
+      return 'userVocabulary'
+    case 'nextUserVocabularyId':
+      return 'userVocabularyNextId'
+    case 'favorites':
+      return 'favorites-2'
+  }
+}
+
+export const getStorageItemOr = async <T,>(key: string, defaultValue: T): Promise<T> => {
+  const value = await AsyncStorage.getItem(key)
+  return value ? JSON.parse(value) : defaultValue
+}
+
+export const getStorageItem = async <T extends keyof Storage>(key: T): Promise<Storage[T]> =>
+  getStorageItemOr(getStorageKey(key), defaultStorage[key])
+
+const setStorageItem = async <T extends keyof Storage>(key: T, value: Storage[T]): Promise<void> => {
+  await AsyncStorage.setItem(getStorageKey(key), JSON.stringify(value))
+}
+
+// https://github.com/react-native-async-storage/async-storage/issues/401#issuecomment-2508924008
+export class StorageCache {
+  private readonly listeners: Map<string, Set<() => void>> = new Map()
```

Comment:
❓ Why did you decide to go for a listener-based approach instead of keeping the storage in a `useState`? Just out of curiosity.

Reply:
The advantage of the listener based approach is that it is more granular. For example, if the storage was kept as a simple state, then setting `storage.foo = new_value()` would not only rerender the current component, but also all other components that read unrelated properties of storage. 
The listener-based approach allows to keep the state updates separate: If i update `storage.foo`, only components that read `storage.foo` will update.